### PR TITLE
Fixed Calculator IA including PercentOf code

### PR DESCRIPTION
**Pull request by puskin94**
Hey, 
I added to the Calculator IA a new rule that allows PercentOf to display result also when input is like 
50% - 50
schermata 2015-01-07 alle 11 14 30
//cc @jagtalon @mintsoft @moollaza

### CODE REVIEW

**Comment by puskin94** [comment]
my fault, I changed only the Calculator.pm file ... Travis is rigth but files about percentof shouldn't be checked because those are just in the PercentOf PR

File: lib/DDG/Goodie/PercentOf.pm [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+		$result = ( $1 - (($1 * $3) / 100) );
+	} elsif ($2 eq '+') {
+		$result = ( $1 + (($1 * $3) / 100) );
+	} elsif ($2 eq '*') {
+		$result = ( $1 * (($1 * $3) / 100) );
+	} elsif ($2 eq '/') {
+		$result = ( $1 / (($1 * $3) / 100) );
+	}
+
+	return "Result: ". $result;
```
Comment (MrChrisW) [inline lib/DDG/Goodie/PercentOf.pm]: Shouldn't this be returning a structured answer like Calculator. 
selection_053

**Comment by MrChrisW** [comment]
@puskin94 Was this PR meant to include the PercentOf files or just the change to Calculator.pm?

Check out the [Using Structured Responses](https://github.com/duckduckgo/duckduckgo-documentation/blob/master/duckduckhack/goodie/goodie_advanced_handle_functions.md)

:smile:

**Comment by puskin94** [comment]
@MrChrisW my fault, this is only a fix for Calculator.pm that allow PercentOf to return a result

**Comment by puskin94** [comment]
Ok ... with https://github.com/duckduckgo/zeroclickinfo-goodies/commit/42e4b456a249d73076c81e85cde642098591fd56 this it should be ok... but for this https://github.com/duckduckgo/p5-app-duckpan/issues/199 bug I can't check it ;( .

**Comment by mintsoft** [comment]
@MrChrisW https://duck.co/duckduckhack/goodie_advanced_handle_functions#using-structured-responses

**Comment by jagtalon** [comment]
@puskin94 nice--it looks like tests are failing though:

```
Test Summary Report                                                                                                                                                          
-------------------                                                                                                                                                          
t/Calculator.t (Wstat: 8448 Tests: 89 Failed: 33)                                                                                                                            
  Failed tests:  1-2, 4, 28-36, 38-40, 42, 45, 50, 53, 58-65                                                                                                                 
                67-70, 73-74                                                                                                                                                 
  Non-zero exit status: 33 
```

when I ran `prove -Ilib t/Calculator.t`

**Comment by puskin94** [comment]
@jagtalon Idk why this `((?<pcnt>\d+)%|(?<num>\d+))(?<op>(+|-|*|/))((?<pcnt>\d+)%|(?<num>\d+))` can broke all the input, also without the percent char! Any tips?

**Comment by mintsoft** [comment]
@puskin94 that regex matches the queries that have failed i.e. `2-2`
```
rob@set $ perl -e 'print "matches\n" if ("2-2" =~ qr/((?<pcnt>\d+)%|(?<num>\d+))(?<op>(\+|\-|\*|\/))((?<pcnt>\d+)%|(?<num>\d+))/);'
matches
```

**Comment by puskin94** [comment]
Hei @mintsoft && @jagtalon I hope I have fixed this little bug but, according to the reason displayed upper, i can't check this out. What About changing this 
```return if ($query =~ qr/((?<pcnt>\d+)%|(?<num>\d+))(?<op>(\+|\-|\*|\/))((?<pcnt>\d+)%|(?<num>\d+))/);```
to this
```return if (($query =~ qr/((?<pcnt>\d+)%|(?<num>\d+))(?<op>(\+|\-|\*|\/))((?<pcnt>\d+)%|(?<num>\d+))/) && ($query =~ qr/%/));
```
??
please try it ;)

**Comment by mintsoft** [comment]
@puskin94 I think the regex is just a little too broad with it's matches. It kind of looks to me like there are 2 cases it's attempting to handle:
`9% + 4` or `4 + 9%`
I'd suggest breaking it up into a regex of each case; something like this:
```perl
qr/
	(?:
		(?<pcnt>\d+)%
		(?<op>(\+|\-|\*|\/))
		(?<num>\d+)
	) |(?:
		(?<num>\d+)
		(?<op>(\+|\-|\*|\/))
		(?<pcnt>\d+)%
	)
/x
```

**Comment by MrChrisW** [comment]
@mintsoft Thanks for the duck.co link, I was being lazy providing the GitHub one :laughing:

Can we make sure this also takes care of the query *what is 10% of 500*

**Comment by puskin94** [comment]
@MrChrisW No problem, when I will fix my duckpan bug, I'll improve this feature in PercentOf PR ;)

**Comment by puskin94** [comment]
@mintsoft @jagtalon I pushed a fix for the regex... Just tell me if It works!
All the Calculator tests are good, and 
```
perl -e 'print "matches\n" if ("2-2" =~ qr/(?:(?<pcnt>\d+)%(?<op>(\+|\-|\*|\/))(?<num>\d+)) | (?:(?<num>\d+)(?<op>(\+|\-|\*|\/))(?<pcnt>\d+)%)/);' 
```
prints nothing!

**Comment by jagtalon** [comment]
@puskin94 nice! What do you think about @mintsoft's suggestion about the regex? https://github.com/duckduckgo/zeroclickinfo-goodies/pull/903#issuecomment-69100812

**Comment by puskin94** [comment]
@jagtalon ?? i already improved that in previous PR ;)

**Comment by jagtalon** [comment]
@puskin94 you mean breaking the regexp into several lines?

**Comment by jagtalon** [comment]
:+1:

**Pull request merged**